src/native/managed/cdac/mscordaccore_universal/Legacy/SOSDacImpl.cs

```diff
@@ -110,8 +110,82 @@ int ISOSDacInterface.GetAppDomainConfigFile(ClrDataAddress appDomain, int count,
 
         return hr;
     }
-    int ISOSDacInterface.GetAppDomainData(ClrDataAddress addr, void* data)
-        => _legacyImpl is not null ? _legacyImpl.GetAppDomainData(addr, data) : HResults.E_NOTIMPL;
+    int ISOSDacInterface.GetAppDomainData(ClrDataAddress addr, DacpAppDomainData* data)
+    {
+        int hr = HResults.S_OK;
+        try
+        {
+            if (addr == 0)
+            {
+                hr = HResults.E_INVALIDARG;
+            }
+            else
+            {
+                NativeMemory.Clear(data, (nuint)sizeof(DacpAppDomainData));
+                data->AppDomainPtr = addr;
+                TargetPointer systemDomainPointer = _target.ReadGlobalPointer(Constants.Globals.SystemDomain);
+                ClrDataAddress systemDomain = _target.ReadPointer(systemDomainPointer).ToClrDataAddress(_target);
+                Contracts.ILoader loader = _target.Contracts.Loader;
+                TargetPointer globalLoaderAllocator = loader.GetGlobalLoaderAllocator();
+                data->pHighFrequencyHeap = loader.GetHighFrequencyHeap(globalLoaderAllocator).ToClrDataAddress(_target);
+                data->pLowFrequencyHeap = loader.GetLowFrequencyHeap(globalLoaderAllocator).ToClrDataAddress(_target);
+                data->pStubHeap = loader.GetStubHeap(globalLoaderAllocator).ToClrDataAddress(_target);
+                data->appDomainStage = (uint)DacpAppDomainDataStage.STAGE_OPEN;
+                if (addr != systemDomain)
+                {
+                    TargetPointer pAppDomain = addr.ToTargetPointer(_target);
+                    data->dwId = _target.ReadGlobal<uint>(Constants.Globals.DefaultADID);
+
+                    List<Contracts.ModuleHandle> modules = loader.GetModuleHandles(
+                        pAppDomain,
+                        AssemblyIterationFlags.IncludeLoading |
+                        AssemblyIterationFlags.IncludeLoaded |
+                        AssemblyIterationFlags.IncludeExecution).ToList();
+
+                    for (int i = 0; i < modules.Count; i++)
+                    {
+                        Contracts.ModuleHandle module = modules[i];
+                        if (loader.IsAssemblyLoaded(module))
+                        {
+                            data->AssemblyCount++;
+                        }
+                    }
+
+                    List<Contracts.ModuleHandle> failedModules = loader.GetModuleHandles(
+                        pAppDomain,
+                        AssemblyIterationFlags.IncludeFailedToLoad).ToList();
+                    data->FailedAssemblyCount = failedModules.Count;
```

Comment:
nit: Can use the `IEnumerable` directly to avoid allocating an array.
```suggestion
                    IEnumerable<Contracts.ModuleHandle> modules = loader.GetModuleHandles(
                        pAppDomain,
                        AssemblyIterationFlags.IncludeLoading |
                        AssemblyIterationFlags.IncludeLoaded |
                        AssemblyIterationFlags.IncludeExecution);
                    foreach (Contracts.ModuleHandle module in modules)
                    {
                        if (loader.IsAssemblyLoaded(module))
                            data->AssemblyCount++;
                    }

                    IEnumerable<Contracts.ModuleHandle> failedModules = loader.GetModuleHandles(
                        pAppDomain,
                        AssemblyIterationFlags.IncludeFailedToLoad);
                    data->FailedAssemblyCount = failedModules.Count();
```